cryptofeed/exchanges/binance.py

```diff
@@ -287,36 +322,140 @@ async def _snapshot(self, pair: str) -> None:
+        self._l2_book[std_pair].sequence_number = resp['lastUpdateId']
+        self._l2_book[std_pair].raw = resp
+        
+        if self.refresh_snapshot:
+            # Delete buffer and reset time since snapshot update
+            self._next_snapshot_time[std_pair] = receipt_timestamp + self._next_snapshot_refresh_interval()
+            self._fetch_snapshop_buffer.pop(std_pair, None)
         ts = self.timestamp_normalize(resp['E']) if 'E' in resp else None
-        await self.book_callback(L2_BOOK, self._l2_book[std_pair], time.time(), raw=resp, timestamp=ts, sequence_number=resp['lastUpdateId'])
+        await self.book_callback(L2_BOOK, self._l2_book[std_pair], receipt_timestamp, timestamp=ts)
+
+    async def _refresh_snapshot(self, pair: str) -> None:
+        resp = await self._fetch_snapshot(pair)
+        receipt_timestamp = time.time()
 
-    async def _handle_book_msg(self, msg: dict, pair: str, timestamp: float):
-        """
-        Processes 'depthUpdate' update book msg. This method should only be called if pair's l2_book exists.
-        """
         std_pair = self.exchange_symbol_to_std_symbol(pair)
-        skip_update = self._check_update_id(std_pair, msg)
-        if skip_update:
+        self.last_update_id[std_pair] = resp['lastUpdateId']
+
+        max_depth = self.max_depth if self.max_depth else self.valid_depths[-1]
+        delta = {BID: [], ASK: []}
+        for key, side in [('bids', BID) , ('asks', ASK)]:
+            for i, (price, amount) in enumerate(resp[key]):
+                price = Decimal(price)
+                amount = Decimal(amount)
+                if price in self._l2_book[std_pair].book[side]:
+                    if amount != self._l2_book[std_pair].book[side][price]:
+                        # Since we haven't looked at all the updates, changes in price are allowed here and will be published as deltas
+                        self._l2_book[std_pair].book[side][price] = amount
+                        if i < max_depth:
+                            delta[side].append((price, amount))
+                else:
+                    # Price not in order book: add it there
+                    self._l2_book[std_pair].book[side][price] = amount
+                    if i < max_depth:
+                        delta[side].append((price, amount))
+        self._l2_book[std_pair].sequence_number = resp['lastUpdateId']
+        
+        if self.refresh_snapshot:
+            # Delete buffer and reset time since snapshot update
+            self._next_snapshot_time[std_pair] = receipt_timestamp + self._next_snapshot_refresh_interval()
+            self._fetch_snapshop_buffer.pop(std_pair, None)
+
+        await self.book_callback(L2_BOOK, self._l2_book[std_pair], receipt_timestamp, delta=delta, sequence_number=self.last_update_id[std_pair])
+        self._l2_book[std_pair].delta = None  # Set delta as None to allow older messages to be skipped correctly
```

Comment:
I'm not sure that setting it after the callback will do anything. What is the purpose of this? It will just be reset at the next callback

Reply:
This method is used in refreshing a snapshot asynchronously. After we have processed this , the next web socket message could have a lastUpdateId <= this ID. This is handled in the `_check_update_id` method, which seemed to have changed in the last few weeks to remove a clause that allowed us to ignore such messages. Without the clause, this case would result in the order book being reset as the system would think it dropped a message. Setting the `delta` as None allows us to fall on the following if statement:
```
if self._l2_book[std_pair].delta is None and msg['u'] <= self.last_update_id[std_pair]:
    return True
```

Tbh it's a hack and probably a better idea would be to add a clause such as:
```
if msg['U'] <= self.last_update_id[std_pair] and msg['u'] <= self.last_update_id[std_pair]:
    return True
```

It feels like if both the start and end IDs are smaller than our last seen ID, we would be able to ignore that message safely. What do you think? I'd prefer removing this line and adding the above check instead.

Reply:
yes that would make more sense and would be easier to read/follow